// simply2d/lib/sdl2int.js
import { createRequire } from 'module';
import * as sdl from "./sdlValues.js";
const require = createRequire(import.meta.url);
let sdl2bind;
try {
    sdl2bind = require("../build/Release/canvas_sdl2.node");
}
catch {
    sdl2bind = require("../build/Debug/canvas_sdl2.node");
}
export { sdl2bind };
let renderingSequence = false;
let currentLayer = null;
export function getTicks() {
    return sdl2bind.getTicks();
}
export function setRenderingSequence() {
    renderingSequence = true;
}
export function clearRenderingSequence() {
    renderingSequence = false;
}
export function getWindow(title, xPos, yPos, width, height, flags) {
    if (sdl2bind.init(sdl.SDL_Init_Everything) != 0) {
        throw "An error occurred while initializing SDL: " + sdl2bind.getError();
    }
    return sdl2bind.createWindow(title, xPos, yPos, width, height, flags);
}
export function getRenderer(window, index, flag) {
    if (window === undefined) {
        throw "Cannot create a window";
    }
    return sdl2bind.createRenderer(window, index, flag);
}
export function showWindow(window) {
    sdl2bind.showWindow(window);
}
export function hideWindow(window) {
    sdl2bind.hideWindow(window);
}
function render(renderer) {
    if (renderingSequence)
        return;
    sdl2bind.renderLayers(renderer);
    sdl2bind.renderPresent(renderer);
}
export function clearWithColor(renderer, r, g, b, alpha) {
    sdl2bind.setDrawColor(renderer, r, g, b, alpha);
    sdl2bind.renderClear(renderer);
    render(renderer);
}
export function getCurrentLayer() {
    currentLayer = sdl2bind.getCurrentLayer();
}
export function delay(ms) {
    sdl2bind.delay(ms);
}
export function setPoint(renderer, r, g, b, a, px, py) {
    sdl2bind.setDrawColor(renderer, r, g, b, a);
    sdl2bind.drawPoint(renderer, px, py);
    render(renderer);
}
export function setLine(renderer, r, g, b, a, px1, py1, px2, py2) {
    sdl2bind.setDrawColor(renderer, r, g, b, a);
    sdl2bind.drawLine(renderer, px1, py1, px2, py2);
    render(renderer);
}
export function setAntialias() {
    sdl2bind.setAntialias();
}
export function setImage(renderer, filename, imgInitFlag) {
    if (sdl2bind.imgInit(imgInitFlag) === 0) {
        throw "An error occurred while setting image: " + sdl2bind.getError();
    }
    let texture = sdl2bind.loadTexture(renderer, filename);
    if (texture === undefined) {
        throw "Cannot load texture from " + filename + ": " + sdl2bind.getError();
    }
    if (sdl2bind.renderCopy(renderer, texture, sdl.NULL, sdl.NULL) !== 0) {
        throw "Cannot load the texture into the renderer";
    }
    render(renderer);
    sdl2bind.imgQuit();
}
export function setPNG(renderer, filename) {
    setImage(renderer, filename, sdl.IMG_Init_Flags.IMG_INIT_PNG);
}
export function setJPG(renderer, filename) {
    setImage(renderer, filename, sdl.IMG_Init_Flags.IMG_INIT_JPG);
}
export function setRectangle(renderer, x, y, width, height, r, g, b, a, fill) {
    sdl2bind.setDrawColor(renderer, r, g, b, a);
    sdl2bind.drawRectangle(renderer, x, y, width, height, fill);
    render(renderer);
}
export function renderPresent(renderer) {
    sdl2bind.renderLayers(renderer);
    sdl2bind.renderPresent(renderer);
}
export function setRawData(renderer, buffer, bitPerPixel, width, height) {
    let pixelFormat = sdl.SDL_PIXEL_FORMAT.SDL_PIXELFORMAT_RGBA8888;
    if (bitPerPixel === 8)
        pixelFormat = sdl.SDL_PIXEL_FORMAT.SDL_PIXELFORMAT_RGB332;
    else if (bitPerPixel === 16)
        pixelFormat = sdl.SDL_PIXEL_FORMAT.SDL_PIXELFORMAT_RGB565;
    else if (bitPerPixel === 24)
        pixelFormat = sdl.SDL_PIXEL_FORMAT.SDL_PIXELFORMAT_RGB888;
    else
        pixelFormat = sdl.SDL_PIXEL_FORMAT.SDL_PIXELFORMAT_RGBA8888;
    let texture = sdl2bind.createTexture(renderer, pixelFormat, sdl.SDL_TEXTURE_ACCESS.SDL_TEXTUREACCESS_STREAMING, width, height);
    sdl2bind.writeTexture(texture, buffer);
    if (sdl2bind.renderCopy(renderer, texture, sdl.NULL, sdl.NULL) !== 0) {
        throw "Cannot load the texture into the renderer";
    }
    ;
    render(renderer);
}
export function watchRawData(renderer, width, height) {
    return new Uint8Array(sdl2bind.readData(renderer, width, height));
}
export function savePNG(renderer, width, height, filename) {
    sdl2bind.savePNG(renderer, width, height, filename);
}
export function saveJPG(renderer, width, height, filename) {
    sdl2bind.saveJPG(renderer, width, height, filename);
}
export function setRenderScale(renderer, width, height, scale) {
    sdl2bind.setScale(renderer, width, height, scale);
}
export function onClickEvent(callback) {
    sdl2bind.onClick(callback);
}
export function onKeyDownEvent(callback) {
    sdl2bind.onKeyDown(callback);
}
export function onKeyUpEvent(callback) {
    sdl2bind.onKeyUp(callback);
}
export function onKeysDownEvent(callback) {
    sdl2bind.onKeysDown(callback);
}
export function onKeysUpEvent(callback) {
    sdl2bind.onKeysUp(callback);
}
export function refresh(renderer) {
    sdl2bind.setDrawColor(renderer, 0, 0, 0, 255);
    sdl2bind.renderClear(renderer);
}
export function setText(renderer, text, r, g, b, x, y) {
    sdl2bind.drawText(renderer, text, r, g, b, x, y);
    render(renderer);
}
export function setArc(renderer, x, y, radius, angle1, angle2, r, g, b, a) {
    sdl2bind.setDrawColor(renderer, r, g, b, a);
    sdl2bind.drawArc(renderer, x, y, radius, angle1, angle2);
    render(renderer);
}
export function setTexture(renderer, x, y, textureFile) {
    sdl2bind.drawTexture(renderer, x, y, textureFile);
    render(renderer);
}
